Mathlib/RingTheory/Polynomial/Hilbert.lean

```diff
@@ -0,0 +1,199 @@
+/-
+Copyright (c) 2024 Fangming Li. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fangming Li, Jujian Zhang
+-/
+import Mathlib.Algebra.Polynomial.RingDivision
+import Mathlib.RingTheory.PowerSeries.WellKnown
+
+/-!
+# Hilbert polynomials
+
+Given any `p : ℤ[X]` and `d : ℕ`, there exists some `h : ℚ[X]` such that for any
+large enough `n : ℕ`, `PowerSeries.coeff ℤ n (p * (@invOneSubPow ℤ _ d))` is equal
+to `h.eval (n : ℚ)`. This `h` is unique and is called the Hilbert polynomial with
+respect to `p` and `d` (`Polynomial.hilbert p d`).
+-/
+
+open BigOperators
+open PowerSeries
+
+namespace Polynomial
+
+section greatestFactorOneSubNotDvd
+
+variable {R : Type*} [CommRing R] [NeZero (1 : R)] [NoZeroDivisors R]
+variable (p : R[X]) (hp : p ≠ 0) (d : ℕ)
+
+/--
+Given a polynomial `p`, the greatest factor of `p` that is not divided by `1 - X`.
+-/
+noncomputable def greatestFactorOneSubNotDvd : R[X] :=
+  ((- 1 : R[X]) ^ p.rootMultiplicity 1) *
+  (exists_eq_pow_rootMultiplicity_mul_and_not_dvd p hp 1).choose
+
+local notation "gFOSND" => greatestFactorOneSubNotDvd
+
+omit [NeZero (1 : R)] [NoZeroDivisors R] in
+theorem pow_rootMultiplicity_mul_greatestFactorOneSubNotDvd_eq :
+    ((1 - X : R[X]) ^ p.rootMultiplicity 1) * greatestFactorOneSubNotDvd p hp = p := by
+  rw [greatestFactorOneSubNotDvd, ← mul_assoc, ← mul_pow]
+  simp only [mul_neg, mul_one, neg_sub, map_one]
+  exact id (exists_eq_pow_rootMultiplicity_mul_and_not_dvd p hp 1).choose_spec.1.symm
+
+omit [NeZero (1 : R)] [NoZeroDivisors R] in
+theorem greatestFactorOneSubNotDvd_ne_zero :
+    greatestFactorOneSubNotDvd p hp ≠ 0 := fun h0 => by
+  let hpow := pow_rootMultiplicity_mul_greatestFactorOneSubNotDvd_eq p hp
+  rw [h0, mul_zero] at hpow; exact hp <| id hpow.symm
+
+theorem natDegree_greatestFactorOneSubNotDvd_le :
+    (greatestFactorOneSubNotDvd p hp).natDegree ≤ p.natDegree := by
+  have : p.natDegree = ((1 - X : R[X]) ^ p.rootMultiplicity 1 * gFOSND p hp).natDegree := by
+    rw [pow_rootMultiplicity_mul_greatestFactorOneSubNotDvd_eq]
+  rw [this, natDegree_mul]
+  exact (gFOSND p hp).natDegree.le_add_left (natDegree ((1 - X) ^ p.rootMultiplicity 1))
+  exact pow_ne_zero _ <| fun h0 => by
+    let this : (1 - X : R[X]).coeff 0 = 0 := by rw [h0]; simp only [coeff_zero];
+    simp only [coeff_sub, coeff_one_zero, coeff_X_zero, sub_zero, one_ne_zero] at this
+  exact greatestFactorOneSubNotDvd_ne_zero p hp
+
+theorem natDegree_pow_rootMultiplicity_sub_mul_greatestFactorOneSubNotDvd_le
+    (hp1 : d + 1 ≤ p.rootMultiplicity 1) :
+    ((1 - X) ^ ((p.rootMultiplicity 1) - (d + 1)) * greatestFactorOneSubNotDvd p hp).natDegree
+    ≤ p.natDegree := by
+  let this := pow_ne_zero (p.rootMultiplicity 1 - (d + 1)) <| fun (h0 : (1 - X : R[X]) = 0) => by
```

Comment:
This just looks pretty much the same as lines 56 to 58. Can you factor out a lemma?